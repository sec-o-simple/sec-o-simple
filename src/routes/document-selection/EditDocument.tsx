import { useSOSImport } from '@/utils/sosDraft'
import { faArrowRight, faEdit } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from '@heroui/button'
import { motion } from 'motion/react'
import { useState } from 'react'
import { useNavigate } from 'react-router'

export default function EditDocument() {
  const navigate = useNavigate()
  const { isSOSDraft, importSOSDocument } = useSOSImport()
  const [jsonObject, setJsonObject] = useState<object | undefined>()

  return (
    <motion.div
      className="flex w-96 flex-col gap-6 rounded-xl border-2 bg-content1 p-6"
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={{ opacity: 0 }}
    >
      <div className="flex items-center gap-2 text-xl font-bold">
        <FontAwesomeIcon className="text-primary" icon={faEdit} />
        Edit existing document
      </div>
      <div>
        <input
          type="file"
          accept=".json,application/json"
          onChange={(e) => {
            if (e.target.files) {
              const file = e.target.files[0]
              const reader = new FileReader()
              reader.onload = function (e) {
                setJsonObject(undefined)
                if (!e.target?.result) {
                  return
                }

                try {
                  const jsonData = JSON.parse(e.target.result as string)
                  setJsonObject(jsonData)
                } catch (err) {
                  console.error('Error parsing JSON:', err)
                }
              }
              reader.readAsText(file)
            }
          }}
          className="w-full rounded-md border p-2 text-sm outline-none focus:border-black"
        />
      </div>
      <div className="self-end">
        <Button
          color="primary"
          endContent={<FontAwesomeIcon icon={faArrowRight} />}
          onPress={() => {
            if (jsonObject) {
              importSOSDocument(jsonObject)
              navigate('/document-information/')
            }
          }}
          isDisabled={!(jsonObject && isSOSDraft(jsonObject))}
        >
          Edit Document
        </Button>
      </div>
    </motion.div>
  )
}
