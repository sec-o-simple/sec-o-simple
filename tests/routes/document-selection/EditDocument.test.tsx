import { describe, it, expect, vi } from 'vitest'
import { render } from '@testing-library/react'
import { HashRouter } from 'react-router'
import EditDocument from '../../../src/routes/document-selection/EditDocument'

// Mock external dependencies  
vi.mock('@heroui/react', () => ({
  Button: ({ children, ...props }: any) => <button {...props}>{children}</button>,
  Modal: ({ children, isOpen, ...props }: any) => isOpen ? <div {...props}>{children}</div> : null,
  ModalContent: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  ModalHeader: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  ModalBody: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  ModalFooter: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  useDisclosure: () => ({
    isOpen: false,
    onOpen: vi.fn(),
    onClose: vi.fn(),
    onOpenChange: vi.fn()
  })
}))

vi.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ icon }: any) => <span data-testid="icon">{typeof icon === 'string' ? icon : 'icon'}</span>
}))

vi.mock('react-router', () => ({
  useNavigate: () => vi.fn(),
  HashRouter: ({ children }: any) => <div>{children}</div>
}))

vi.mock('../../../src/utils/useDocumentStore', () => ({
  default: () => ({
    resetDocument: vi.fn(),
    setDocument: vi.fn()
  })
}))

vi.mock('../../../src/utils/useDatabaseClient', () => ({
  default: () => ({
    setDocument: vi.fn()
  })
}))

vi.mock('../../../src/utils/csafImport/parseCSAFDocument', () => ({
  default: vi.fn()
}))

vi.mock('../../../src/utils/csafImport/useCSAFImport', () => ({
  default: () => ({
    importDocument: vi.fn(),
    hasHiddenFields: false
  })
}))

vi.mock('../../../src/utils/sosDraft', () => ({
  isDraftDocument: vi.fn(),
  useSOSImport: () => ({
    isSOSDraft: vi.fn(),
    importSOSDocument: vi.fn()
  })
}))

describe('EditDocument', () => {
  const renderComponent = () => {
    return render(
      <HashRouter>
        <EditDocument />
      </HashRouter>
    )
  }

  it('should render correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })
})
