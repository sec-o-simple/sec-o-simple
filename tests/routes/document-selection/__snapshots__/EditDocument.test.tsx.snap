// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditDocument > should render correctly 1`] = `
<div>
  <div>
    <div
      class="flex w-96 flex-col gap-6 rounded-xl border-2 bg-content1 p-6"
      style="opacity: 0;"
    >
      <div
        class="flex items-center gap-2 text-xl font-bold"
      >
        <span
          data-testid="icon"
        >
          icon
        </span>
        Edit existing document
      </div>
      <div>
        <input
          accept=".json,application/json"
          class="w-full rounded-md border p-2 text-sm outline-none focus:border-black"
          type="file"
        />
      </div>
      <div
        class="self-end"
      >
        <button
          data-color="primary"
          data-isdisabled="true"
          data-isicononly="false"
          data-isinvalid="false"
          data-testid="hero-button"
          disabled=""
          endcontent="[object Object]"
        >
          Edit Document
        </button>
      </div>
    </div>
  </div>
</div>
`;
